language/stdlib/modules/transaction_fee_distribution.mvir

```diff
@@ -0,0 +1,100 @@
+// This module implements a basic transaction fee distribution logic.
+//
+// Current limitations are:
+// * We don't pay out per-epoch, and instead pay-out per-block.
+// * In the handling of dust, it could be beneficial to be the first validator in the validating
+//   set. However, the maximal benefit that can be achieved per-block(/epoch) is at most 1 microlibra.
+module TransactionFeeDistribution {
+  import 0x0.LibraCoin;
+  import 0x0.Block;
+  import 0x0.ValidatorSet;
+  import 0x0.LibraAccount;
+  import 0x0.TransactionFeeCollection;
+
+   public distribute_transaction_fees() {
```

Comment:
If block metadata transactions can call private functions, then this would be the way to go. I'm just thinking of in the meantime we probably want to have this public so we can write tests against them. But long-term, totally agreed.

Reply:
Makes sense!

Reply:
Yes! The block metadata transaction WILL be able to call to the private functions. IMHO that's the only way for us to prevent code from being called by arbitrary person.

Reply:
> Yes! The block metadata transaction WILL be able to call to the private functions. IMHO that's the only way for us to prevent code from being called by arbitrary person.

I'm not sure its the "only" way. Couldn't we set the sender to the association or node or something, and then check the ValidatorSet? 

Additionally, its not a thing I like too much, but we could also add `is_metadata_transaction()`. It's not great. But to me it feels less gross than the VM calling private functions

Reply:
I think in general we need to have a better story around metadata transactions since we would want to probably statically disallow such things as  `move_to_sender` since it makes no sense in such cases.  

I think we should break this discussion out, since I don't think we'll solve it in this PR, but I think it's an important design aspect that needs to be discussed ASAP.  One proposal I would toss out as a consideration; have different function modifiers such as `public`, `metadata` etc. These could then syntactically restrict allowable forms in the function body.